File: Medium/2026-08-09-1140-Stone-Game-II/solution.java
class Solution {
    public int stoneGameII(int[] piles) {
        int n = piles.length;
        int[][] dp = new int[n][n + 1];
        int[] prefixSum = new int[n + 1];

        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + piles[i];
        }

        return dfs(0, 1, piles, dp, prefixSum);
    }

    private int dfs(int index, int M, int[] piles, int[][] dp, int[] prefixSum) {
        if (index >= piles.length) return 0;
        if (dp[index][M] != 0) return dp[index][M];

        int maxStones = 0;
        for (int x = 1; x <= 2 * M; x++) {
            if (index + x > piles.length) break;
            int stones = prefixSum[index + x] - prefixSum[index];
            maxStones = Math.max(maxStones, stones + prefixSum[piles.length] - prefixSum[index + x] - dfs(index + x, Math.max(M, x), piles, dp, prefixSum));
        }

        dp[index][M] = maxStones;
        return maxStones;
    }
}

File: Medium/2026-08-09-1140-Stone-Game-II/solution.js
var stoneGameII = function(piles) {
    const n = piles.length;
    const dp = Array.from({ length: n + 1 }, () => Array(n + 1).fill(0));
    const prefixSum = new Array(n + 1).fill(0);

    for (let i = 0; i < n; i++) {
        prefixSum[i + 1] = prefixSum[i] + piles[i];
    }

    const dfs = (i, m) => {
        if (i >= n) return 0;
        if (dp[i][m] !== 0) return dp[i][m];

        let maxStones = 0;
        for (let x = 1; x <= 2 * m; x++) {
            if (i + x > n) break;
            const stones = prefixSum[i + x] - prefixSum[i];
            maxStones = Math.max(maxStones, stones - dfs(i + x, Math.max(m, x)));
        }

        dp[i][m] = maxStones;
        return maxStones;
    };

    return dfs(0, 1);
};